tox.ini

```diff
@@ -76,6 +76,7 @@ deps =
   opentelemetry: pytest
   opentelemetry: pytest-benchmark
   opentelemetry: flaky
+  opentelemetry-exporter-otlp-proto-http: responses
```

Comment:
I think this can go here https://github.com/open-telemetry/opentelemetry-python/blob/main/exporter/opentelemetry-exporter-otlp-proto-http/pyproject.toml#L37

Reply:
Hmmmm. If I remove this from here and from `dev-requirements.txt` and add it to `pyproject.toml` the tests fail locally because the package is missing:

```diff
diff --git a/dev-requirements.txt b/dev-requirements.txt
index f930e85e8..e3fced6b8 100644
--- a/dev-requirements.txt
+++ b/dev-requirements.txt
@@ -19,4 +19,3 @@ protobuf~=3.18.1
 markupsafe==2.0.1
 bleach==4.1.0 # This dependency was updated to a breaking version.
 codespell==2.1.0
-responses==0.22.0
diff --git a/exporter/opentelemetry-exporter-otlp-proto-http/pyproject.toml b/exporter/opentelemetry-exporter-otlp-proto-http/pyproject.toml
index edfc307c1..a89e7fa18 100644
--- a/exporter/opentelemetry-exporter-otlp-proto-http/pyproject.toml
+++ b/exporter/opentelemetry-exporter-otlp-proto-http/pyproject.toml
@@ -34,7 +34,9 @@ dependencies = [
 ]
 
 [project.optional-dependencies]
-test = []
+test = [
+  "responses == 0.22.0",
+]
 
 [project.entry-points.opentelemetry_traces_exporter]
 otlp_proto_http = "opentelemetry.exporter.otlp.proto.http.trace_exporter:OTLPSpanExporter"
diff --git a/tox.ini b/tox.ini
index 7f647b3dd..1a24186fb 100644
--- a/tox.ini
+++ b/tox.ini
@@ -76,7 +76,6 @@ deps =
   opentelemetry: pytest
   opentelemetry: pytest-benchmark
   opentelemetry: flaky
-  opentelemetry-exporter-otlp-proto-http: responses
   coverage: pytest
   coverage: pytest-cov
   mypy,mypyinstalled: mypy
```

<details>
<summary>Full example output</summary>

```
$ tox -f py310-opentelemetry-exporter-otlp-proto-http
py310-opentelemetry-exporter-otlp-proto-http create: /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http
py310-opentelemetry-exporter-otlp-proto-http installdeps: -cdev-requirements.txt, pytest, pytest-benchmark, flaky
py310-opentelemetry-exporter-otlp-proto-http installed: attrs==22.1.0,flaky==3.7.0,iniconfig==1.1.1,packaging==21.3,pluggy==1.0.0,py==1.11.0,py-cpuinfo==9.0.0,pyparsing==3.0.9,pytest==7.1.3,pytest-benchmark==4.0.0,tomli==2.0.1
py310-opentelemetry-exporter-otlp-proto-http run-test-pre: PYTHONHASHSEED='2007173679'
py310-opentelemetry-exporter-otlp-proto-http run-test-pre: commands[0] | python -m pip install -U pip setuptools wheel
Requirement already satisfied: pip in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (22.3)
Requirement already satisfied: setuptools in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (65.5.0)
Requirement already satisfied: wheel in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (0.37.1)
py310-opentelemetry-exporter-otlp-proto-http run-test-pre: commands[1] | pip install /Users/nst/src/opentelemetry-python/opentelemetry-api /Users/nst/src/opentelemetry-python/opentelemetry-semantic-conventions /Users/nst/src/opentelemetry-python/opentelemetry-sdk /Users/nst/src/opentelemetry-python/tests/opentelemetry-test-utils
Processing /Users/nst/src/opentelemetry-python/opentelemetry-api
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Processing /Users/nst/src/opentelemetry-python/opentelemetry-semantic-conventions
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Processing /Users/nst/src/opentelemetry-python/opentelemetry-sdk
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Processing /Users/nst/src/opentelemetry-python/tests/opentelemetry-test-utils
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: setuptools>=16.0 in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (from opentelemetry-api==1.13.0) (65.5.0)
Collecting deprecated>=1.2.6
  Using cached Deprecated-1.2.13-py2.py3-none-any.whl (9.6 kB)
Collecting typing-extensions>=3.7.4
  Using cached typing_extensions-4.4.0-py3-none-any.whl (26 kB)
Collecting asgiref~=3.0
  Using cached asgiref-3.5.2-py3-none-any.whl (22 kB)
Collecting wrapt<2,>=1.10
  Using cached wrapt-1.14.1-cp310-cp310-macosx_11_0_arm64.whl (35 kB)
Building wheels for collected packages: opentelemetry-api, opentelemetry-semantic-conventions, opentelemetry-sdk, opentelemetry-test-utils
  Building wheel for opentelemetry-api (pyproject.toml) ... done
  Created wheel for opentelemetry-api: filename=opentelemetry_api-1.13.0-py3-none-any.whl size=50942 sha256=1901355eef56579ac74db918b4516c56279224c44c0e60e649dc64d9ef7e5c74
  Stored in directory: /Users/nst/Library/Caches/pip/wheels/5d/53/d9/f0d9caa87536842a7e47fad56709c457f5079605c9e9cffd07
  Building wheel for opentelemetry-semantic-conventions (pyproject.toml) ... done
  Created wheel for opentelemetry-semantic-conventions: filename=opentelemetry_semantic_conventions-0.34b0-py3-none-any.whl size=26484 sha256=f7173b52a2b82a26871000ad66a7fc4d924bd442bedc7cda7d8000483944437a
  Stored in directory: /Users/nst/Library/Caches/pip/wheels/a3/64/be/34a26e1a995c689302bf594158f2d0bdcea277cf723b2518b6
  Building wheel for opentelemetry-sdk (pyproject.toml) ... done
  Created wheel for opentelemetry-sdk: filename=opentelemetry_sdk-1.13.0-py3-none-any.whl size=82802 sha256=f772045730af6fca5b3b1c163be475730f788d4a45a1529b68146cea95f77885
  Stored in directory: /Users/nst/Library/Caches/pip/wheels/0a/e4/0f/cc3c330cc9a992f48d3a13f4aae80992e880b548ba02426685
  Building wheel for opentelemetry-test-utils (pyproject.toml) ... done
  Created wheel for opentelemetry-test-utils: filename=opentelemetry_test_utils-0.34b0-py3-none-any.whl size=12326 sha256=22a289670a4d4f9ebce121ec0646b0f9e5f0baf7ba5da37d62afa36ed1cb3f20
  Stored in directory: /Users/nst/Library/Caches/pip/wheels/eb/29/94/cd64198ec54c8d7cd3e8c0768a8fc3063e8282d6f20c8093fa
Successfully built opentelemetry-api opentelemetry-semantic-conventions opentelemetry-sdk opentelemetry-test-utils
Installing collected packages: wrapt, typing-extensions, opentelemetry-semantic-conventions, asgiref, deprecated, opentelemetry-api, opentelemetry-sdk, opentelemetry-test-utils
Successfully installed asgiref-3.5.2 deprecated-1.2.13 opentelemetry-api-1.13.0 opentelemetry-sdk-1.13.0 opentelemetry-semantic-conventions-0.34b0 opentelemetry-test-utils-0.34b0 typing-extensions-4.4.0 wrapt-1.14.1
py310-opentelemetry-exporter-otlp-proto-http run-test-pre: commands[2] | pip install /Users/nst/src/opentelemetry-python/opentelemetry-proto
Processing /Users/nst/src/opentelemetry-python/opentelemetry-proto
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting protobuf~=3.13
  Using cached protobuf-3.20.3-py2.py3-none-any.whl (162 kB)
Building wheels for collected packages: opentelemetry-proto
  Building wheel for opentelemetry-proto (pyproject.toml) ... done
  Created wheel for opentelemetry-proto: filename=opentelemetry_proto-1.13.0-py3-none-any.whl size=66508 sha256=093f35e98d867398f23f13c5f7be70aa644489d36c1d1b2cb31ecb3bdde01c0d
  Stored in directory: /Users/nst/Library/Caches/pip/wheels/78/d7/a4/394b9446656f9ef5312ac25571eef6f449571d8fd4e9346c05
Successfully built opentelemetry-proto
Installing collected packages: protobuf, opentelemetry-proto
Successfully installed opentelemetry-proto-1.13.0 protobuf-3.20.3
py310-opentelemetry-exporter-otlp-proto-http run-test-pre: commands[3] | pip install /Users/nst/src/opentelemetry-python/exporter/opentelemetry-exporter-otlp-proto-http
Processing /Users/nst/src/opentelemetry-python/exporter/opentelemetry-exporter-otlp-proto-http
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: opentelemetry-sdk~=1.12 in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (from opentelemetry-exporter-otlp-proto-http==1.13.0) (1.13.0)
Requirement already satisfied: opentelemetry-proto==1.13.0 in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (from opentelemetry-exporter-otlp-proto-http==1.13.0) (1.13.0)
Requirement already satisfied: opentelemetry-api~=1.12 in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (from opentelemetry-exporter-otlp-proto-http==1.13.0) (1.13.0)
Collecting googleapis-common-protos~=1.52
  Using cached googleapis_common_protos-1.56.4-py2.py3-none-any.whl (211 kB)
Collecting backoff<3.0.0,>=1.10.0
  Using cached backoff-2.2.1-py3-none-any.whl (15 kB)
Collecting requests~=2.7
  Using cached requests-2.28.1-py3-none-any.whl (62 kB)
Requirement already satisfied: protobuf~=3.13 in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (from opentelemetry-proto==1.13.0->opentelemetry-exporter-otlp-proto-http==1.13.0) (3.20.3)
Requirement already satisfied: setuptools>=16.0 in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (from opentelemetry-api~=1.12->opentelemetry-exporter-otlp-proto-http==1.13.0) (65.5.0)
Requirement already satisfied: deprecated>=1.2.6 in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (from opentelemetry-api~=1.12->opentelemetry-exporter-otlp-proto-http==1.13.0) (1.2.13)
Requirement already satisfied: opentelemetry-semantic-conventions==0.34b0 in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (from opentelemetry-sdk~=1.12->opentelemetry-exporter-otlp-proto-http==1.13.0) (0.34b0)
Requirement already satisfied: typing-extensions>=3.7.4 in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (from opentelemetry-sdk~=1.12->opentelemetry-exporter-otlp-proto-http==1.13.0) (4.4.0)
Collecting urllib3<1.27,>=1.21.1
  Using cached urllib3-1.26.12-py2.py3-none-any.whl (140 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2022.9.24-py3-none-any.whl (161 kB)
Collecting charset-normalizer<3,>=2
  Using cached charset_normalizer-2.1.1-py3-none-any.whl (39 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.4-py3-none-any.whl (61 kB)
Requirement already satisfied: wrapt<2,>=1.10 in /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/lib/python3.10/site-packages (from deprecated>=1.2.6->opentelemetry-api~=1.12->opentelemetry-exporter-otlp-proto-http==1.13.0) (1.14.1)
Building wheels for collected packages: opentelemetry-exporter-otlp-proto-http
  Building wheel for opentelemetry-exporter-otlp-proto-http (pyproject.toml) ... done
  Created wheel for opentelemetry-exporter-otlp-proto-http: filename=opentelemetry_exporter_otlp_proto_http-1.13.0-py3-none-any.whl size=21546 sha256=f911e0b13e6308bf8ffef25c3a104c3d49ce768913c2c266d634d335d17628de
  Stored in directory: /Users/nst/Library/Caches/pip/wheels/1b/09/ea/048934d4442efb8b657f8e77ebfd0e63302091d865d6d8de5c
Successfully built opentelemetry-exporter-otlp-proto-http
Installing collected packages: urllib3, idna, googleapis-common-protos, charset-normalizer, certifi, backoff, requests, opentelemetry-exporter-otlp-proto-http
Successfully installed backoff-2.2.1 certifi-2022.9.24 charset-normalizer-2.1.1 googleapis-common-protos-1.56.4 idna-3.4 opentelemetry-exporter-otlp-proto-http-1.13.0 requests-2.28.1 urllib3-1.26.12
py310-opentelemetry-exporter-otlp-proto-http run-test: commands[0] | pytest
=========================================================================================================================================================================================== test session starts ============================================================================================================================================================================================
platform darwin -- Python 3.10.8, pytest-7.1.3, pluggy-1.0.0 -- /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/bin/python
cachedir: .tox/py310-opentelemetry-exporter-otlp-proto-http/.pytest_cache
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/nst/src/opentelemetry-python, configfile: pyproject.toml
plugins: flaky-3.7.0, benchmark-4.0.0
collected 3 items / 3 errors

================================================================================================================================================================================================== ERRORS ==================================================================================================================================================================================================
____________________________________________________________________________________________________________________________________________________ ERROR collecting exporter/opentelemetry-exporter-otlp-proto-http/tests/test_proto_log_exporter.py _____________________________________________________________________________________________________________________________________________________
ImportError while importing test module '/Users/nst/src/opentelemetry-python/exporter/opentelemetry-exporter-otlp-proto-http/tests/test_proto_log_exporter.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test_proto_log_exporter.py:22: in <module>
    import responses
E   ModuleNotFoundError: No module named 'responses'
____________________________________________________________________________________________________________________________________________________ ERROR collecting exporter/opentelemetry-exporter-otlp-proto-http/tests/test_proto_span_exporter.py ____________________________________________________________________________________________________________________________________________________
ImportError while importing test module '/Users/nst/src/opentelemetry-python/exporter/opentelemetry-exporter-otlp-proto-http/tests/test_proto_span_exporter.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test_proto_span_exporter.py:20: in <module>
    import responses
E   ModuleNotFoundError: No module named 'responses'
_______________________________________________________________________________________________________________________________________________ ERROR collecting exporter/opentelemetry-exporter-otlp-proto-http/tests/metrics/test_otlp_metrics_exporter.py _______________________________________________________________________________________________________________________________________________
ImportError while importing test module '/Users/nst/src/opentelemetry-python/exporter/opentelemetry-exporter-otlp-proto-http/tests/metrics/test_otlp_metrics_exporter.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
metrics/test_otlp_metrics_exporter.py:19: in <module>
    import responses
E   ModuleNotFoundError: No module named 'responses'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================================================================================================================================ 3 errors in 0.23s =============================================================================================================================================================================================
ERROR: InvocationError for command /Users/nst/src/opentelemetry-python/.tox/py310-opentelemetry-exporter-otlp-proto-http/bin/pytest (exited with code 2)
_________________________________________________________________________________________________________________________________________________________________________________________________ summary __________________________________________________________________________________________________________________________________________________________________________________________________
ERROR:   py310-opentelemetry-exporter-otlp-proto-http: commands failed
tox -f py310-opentelemetry-exporter-otlp-proto-http  6.28s user 1.31s system 69% cpu 84944M memory 10.954 total
```
</details>

Any ideas what we're missing?

Reply:
I thought it installed these optional dependencies with `[test]` but it seems that's not the case here. You can ignore my comment.

Reply:
There's definitely still something wrong about how we're doing this because the linter's complaining about it too...

```
************* Module exporter.opentelemetry-exporter-otlp-proto-http.tests.test_proto_span_exporter
exporter/opentelemetry-exporter-otlp-proto-http/tests/test_proto_span_exporter.py:20:0: E0401: Unable to import 'responses' (import-error)
exporter/opentelemetry-exporter-otlp-proto-http/tests/test_proto_span_exporter.py:198:4: R0201: Method could be a function (no-self-use)
************* Module exporter.opentelemetry-exporter-otlp-proto-http.tests.test_proto_log_exporter
exporter/opentelemetry-exporter-otlp-proto-http/tests/test_proto_log_exporter.py:22:0: E0401: Unable to import 'responses' (import-error)
************* Module exporter.opentelemetry-exporter-otlp-proto-http.tests.metrics.test_otlp_metrics_exporter
exporter/opentelemetry-exporter-otlp-proto-http/tests/metrics/test_otlp_metrics_exporter.py:19:0: E0401: Unable to import 'responses' (import-error)
```

Reply:
Pushed another change which does what you suggested _and_ makes sure the `test` extra is installed while running the tests. Hopefully that satisfies both the tests and the linter.

Reply:
Thank you, it will be all 🟢 this time. The `[test]` should have been there already. I will create a separate issue for myself updating tooling stuff.